### DISCUSSION TITLE
Use variables in Timeout syntax on Github action pipeline

### DISCUSSION THREAD

**Discussion by obelisk0114** [community/community#26332] [Actions]
<p>I want to use some variables to set values in GitHub actions steps.</p>
<p><a href="https://docs.github.com/en/actions/learn-github-actions/environment-variables" rel="noopener nofollow ugc">Environment variables</a> tells us that we can set custom environment variables in your workflow file.</p>
<blockquote>
<p>By default, Linux runners use the bash shell, so you must use the syntax <code>$NAME</code> .</p>
</blockquote>
<p>So I try</p>
<pre><code class="lang-auto">name: Workflow

on:
  push:
    branches:
    - main
  pull_request:

env:
  General_Timeout_Minutes: 15

jobs:
  test:
    name: Test
    runs-on: ubuntu-latest
    env:
      Setup_Timeout_Minutes: 10
      Lint_Timeout_Minutes: $General_Timeout_Minutes
      Unit_Test_Timeout_Minutes: $General_Timeout_Minutes
      E2E_Timeout_Minutes: $General_Timeout_Minutes
    steps:
      - uses: actions/checkout@v2

      - name: Setup
        timeout-minutes: $Setup_Timeout_Minutes
        run: npm install

      - name: Lint JavaScript
        timeout-minutes: $Lint_Timeout_Minutes
        run: npm run lint

      - name: Run JavaScript tests
        timeout-minutes: $Unit_Test_Timeout_Minutes
        run: npm run unit-test

      - name: Run end-to-end browser tests
        timeout-minutes: $E2E_Timeout_Minutes
        run: npm run e2e-test

  deploy:
    ...
</code></pre>
<p>However, GitHub actions starts fail with</p>
<blockquote>
<p>The workflow is not valid. .github/workflows/test.yml: Unexpected value ‘$Setup_Timeout_Minutes’ .github/workflows/test.yml: Unexpected value ‘$Lint_Timeout_Minutes’</p>
</blockquote>
<p>How to use variables in <code>timeout-minutes</code> syntax?</p>

**Comment by jsoref** [comment]
<ol>
<li>
<code>$foo</code> is a shell variable, generally bash.</li>
<li>Most of workflows use <code>${{ ... }}</code> to access objects, they aren’t bash.</li>
<li>You can’t forward reference variables. A variable you’re referencing must exist when you reference it if you want to get a non empty value.</li>
<li>You should read about contexts.</li>
</ol>

**Comment by airtower-luna** [comment]
<p>The only part of a workflow that gets run through a shell (which on Linux is Bash by default) is the content of a <code>run</code> step. In there you should use Bash syntax. In other parts of the workflow you need to use the workflow variable substitution (with <code>${{ ... }}</code>). And mind that <code>env</code> variables aren’t available everywhere, see <a href="https://docs.github.com/en/actions/learn-github-actions/contexts#env-context" class="inline-onebox">Contexts - GitHub Docs</a>.</p>

**Comment by obelisk0114** [comment]
<blockquote>
<pre><code>env:
 Setup_Timeout_Minutes: 10
 Lint_Timeout_Minutes: $General_Timeout_Minutes
 Unit_Test_Timeout_Minutes: $General_Timeout_Minutes
 E2E_Timeout_Minutes: $General_Timeout_Minutes
</code></pre>
</blockquote>
<p>If I change <code>env</code> in the <code>jobs</code> as</p>
<pre><code class="lang-auto">    env:
      Setup_Timeout_Minutes: 10
      Lint_Timeout_Minutes: ${{ env.General_Timeout_Minutes }}
      Unit_Test_Timeout_Minutes: ${{ env.General_Timeout_Minutes }}
      E2E_Timeout_Minutes: ${{ env.General_Timeout_Minutes }}
</code></pre>
<p>and change every <code>timeout-minutes</code> to</p>
<pre><code class="lang-auto">timeout-minutes: ${{ env.Lint_Timeout_Minutes }}
</code></pre>
<p>It will show</p>
<blockquote>
<p>Unrecognized named-value: ‘env’. In Lint_Timeout_Minutes: ${{ env.General_Timeout_Minutes }}</p>
</blockquote>
<p>I get another error in <code>timeout-minutes: ${{ env.Setup_Timeout_Minutes }}</code>.</p>
<blockquote>
<p>Unexpected value ‘10’</p>
</blockquote>

**Comment by airtower-luna** [comment]
<p>The syntax is actions syntax now, but the <code>env</code> context isn’t available in the <code>env</code> setting of a job. As described in the documentation I linked above you can only use it within steps:</p>
<blockquote>
<p>You can use the <code>env</code> context in the value of any key in a step except for the <code>id</code> and <code>uses</code> keys.</p>
</blockquote>
<p>The easiest thing to do would be to define your environment variables directly at the job level.</p>
<aside class="quote no-group quote-modified" data-username="obelisk0114" data-post="4" data-topic="252149">
<div class="title">
<div class="quote-controls"></div>
 obelisk0114:</div>
<blockquote>
<p>I get another error in <code>timeout-minutes: ${{ env.Setup_Timeout_Minutes }}</code> .</p>
<blockquote>
<p>Unexpected value ‘10’</p>
</blockquote>
</blockquote>
</aside>
<p>That looks like <code>timeout-minutes</code> expects a number, but gets a string, because that’s what an environment variable is. Maybe <a href="https://docs.github.com/en/actions/learn-github-actions/expressions#fromjson"><code>fromJSON()</code></a> can help there (I haven’t tried), but it might be much simpler to write a number.</p>

**Comment by obelisk0114** [comment, marked as answer]
<aside class="quote no-group">
<blockquote>
<p>The easiest thing to do would be to define your environment variables directly at the job level.</p>
</blockquote>
</aside>
<p>I set them in the job successfully. This is the code snippets.</p>
<pre><code class="lang-auto">run: |
  echo "Lint_Timeout_Minutes=$General_Timeout_Minutes" &gt;&gt; $GITHUB_ENV
  echo "Unit_Test_Timeout_Minutes=$General_Timeout_Minutes" &gt;&gt; $GITHUB_ENV
  echo "E2E_Timeout_Minutes=$General_Timeout_Minutes" &gt;&gt; $GITHUB_ENV
</code></pre>
<aside class="quote group-Advisory-Council" data-username="airtower-luna" data-post="5" data-topic="252149">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>Maybe <a href="https://docs.github.com/en/actions/learn-github-actions/expressions#fromjson" rel="noopener nofollow ugc"> <code>fromJSON()</code> </a> can help there</p>
</blockquote>
</aside>
<p>Thanks for pointing out the road. I also find an <a href="https://docs.github.com/en/actions/learn-github-actions/expressions#example-returning-a-json-data-type" rel="noopener nofollow ugc">example returning a JSON data type</a></p>
<pre><code class="lang-auto">env: 
  continue: true
  time: 3
jobs:
  job1:
    runs-on: ubuntu-latest
    steps:
      - continue-on-error: ${{ fromJSON(env.continue) }}
        timeout-minutes: ${{ fromJSON(env.time) }}
        run: echo ...
</code></pre>
<p>Here is the working one.</p>
<pre><code class="lang-auto">name: Workflow

on:
  push:
    branches:
    - main
  pull_request:

env:
  General_Timeout_Minutes: 15

jobs:
  test:
    name: Test
    runs-on: ubuntu-latest
    env:
      Setup_Timeout_Minutes: 10
    steps:
      - uses: actions/checkout@v2

      # Set job level environment variables from global ones
      - name: "Set job env from global env"
        run: |
          echo "Lint_Timeout_Minutes=$General_Timeout_Minutes" &gt;&gt; $GITHUB_ENV
          echo "Unit_Test_Timeout_Minutes=$General_Timeout_Minutes" &gt;&gt; $GITHUB_ENV
          echo "E2E_Timeout_Minutes=$General_Timeout_Minutes" &gt;&gt; $GITHUB_ENV

      - name: Setup
        timeout-minutes: ${{ fromJSON(env.Setup_Timeout_Minutes) }}
        run: npm install

      - name: Lint JavaScript
        timeout-minutes: ${{ fromJSON(env.Lint_Timeout_Minutes) }}
        run: npm run lint

      - name: Run JavaScript tests
        timeout-minutes: ${{ fromJSON(env.Unit_Test_Timeout_Minutes) }}
        run: npm run unit-test

      - name: Run end-to-end browser tests
        timeout-minutes: ${{ fromJSON(env.E2E_Timeout_Minutes) }}
        run: npm run e2e-test

  deploy:
    ...
     
</code></pre>
